### PR TITLE
Intel fpga constant type inference

### PR DESCRIPTION
**Pull request by TizianoDeMatteis**
- Solves problem with type inference for constants
- reduces length of  function names

### CODE REVIEW

File: tests/intel_fpga_test.sh [hunk 1]
```diff
@@ -130,6 +130,10 @@ run_all() {
 
     # Channels mangling
     run_sample intel_fpga/channels_mangling channels_mangling "\n"
+
+    #Constant Type inference
```
Comment (tbennun) [inline tests/intel_fpga_test.sh:134]:
```suggestion
    # Constant Type inference
```

File: tests/type_inference_test.py [hunk 2]
```diff
@@ -97,6 +97,12 @@ def testExpressionAssignment(self):
...
         self.assertEqual(symbols["res3"], dtypes.typeclass(int))
 
+    def testArrayAccess(self):
+
+        code_str = "tmp = array[i]"
+        symbols = type_inference.infer_types(code_str, {"array": dtypes.typeclass(float)})
+        self.assertEqual(symbols["tmp"], dtypes.typeclass(float))
+
     def testAssignmentIf(self):
         code_str = "res = 5 if(x>10) else 3.1"
```
Comment (tbennun) [inline tests/type_inference_test.py:107]:
```suggestion
        code_str = "res = 5 if x > 10 else 3.1"
```

File: tests/intel_fpga/constant_type_inference.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    fpga_state.add_memlet_path(map_entry, tasklet, memlet=dace.Memlet())
+    fpga_state.add_memlet_path(tasklet,
+                               map_exit,
+                               out,
+                               src_conn="out",
+                               memlet=dace.Memlet("device_output[i]"))
+
+    sdfg.add_edge(fpga_state, copy_out_state, dace.sdfg.sdfg.InterstateEdge())
+    sdfg.fill_scope_connectors()
+    sdfg.save("/tmp/out.sdfg")
```
Comment (tbennun) [inline tests/intel_fpga/constant_type_inference.py:63]:
```suggestion
```

File: tests/type_inference_test.py [hunk 4]
```diff
@@ -97,6 +97,12 @@ def testExpressionAssignment(self):
         symbols = type_inference.infer_types(code_str, symbols)
         self.assertEqual(symbols["res3"], dtypes.typeclass(int))
 
+    def testArrayAccess(self):
+
```
Comment (tbennun) [inline tests/type_inference_test.py:101]:
```suggestion
```

**Review by tbennun — CHANGES_REQUESTED** [review]
minor changes

File: dace/codegen/targets/intel_fpga.py [hunk 5]
```diff
@@ -1326,13 +1345,34 @@ def __init__(self, sdfg, defined_vars, memlets, codegen):
         # Keep track of the different streams used in a tasklet
         self.used_streams = []
         self.width_converters = set()  # Pack and unpack vectors
-        self.dtypes = {k: v[3] for k, v in memlets.items()}  # Type inference
+        self.dtypes = {k: v[3]
+                       for k, v in memlets.items()
+                       if k is not None}  # Type inference
+        # consider also constants: add them to known dtypes
+        for k, v in sdfg.constants.items():
+            if k is not None:
```
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply dace/codegen/targets/intel_fpga.py:1353]: It could be None if you have empty memlets

**Review by tbennun — APPROVED** [review]

**Pull request merged**